lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -1723,6 +1763,177 @@ class SCFToCalyxPass : public circt::impl::SCFToCalyxBase<SCFToCalyxPass> {
+        callee.getFunctionType().getInputs());
+    updatedCalleeArgTypes.append(extraMemRefArgTypes.begin(),
+                                 extraMemRefArgTypes.end());
+    callee.setType(FunctionType::get(callee.getContext(), updatedCalleeArgTypes,
+                                     callee.getFunctionType().getResults()));
+
+    unsigned otherArgsCount = 0;
+    SmallVector<Value, 4> calleeArgFnOperands;
+    builder.setInsertionPointToStart(callerEntryBlock);
+    for (auto arg : callee.getArguments().take_front(originalCalleeArgNum)) {
+      if (isa<MemRefType>(arg.getType())) {
+        auto memrefType = cast<MemRefType>(arg.getType());
+        auto allocOp =
+            builder.create<memref::AllocOp>(callee.getLoc(), memrefType);
+        calleeArgFnOperands.push_back(allocOp);
+      } else {
+        auto callerArg = callerEntryBlock->getArgument(otherArgsCount++);
+        calleeArgFnOperands.push_back(callerArg);
+      }
+    }
+
+    SmallVector<Value, 4> fnOperands;
+    fnOperands.append(calleeArgFnOperands.begin(), calleeArgFnOperands.end());
+    fnOperands.append(extraMemRefOperands.begin(), extraMemRefOperands.end());
+    auto calleeName =
+        SymbolRefAttr::get(builder.getContext(), callee.getSymName());
+    auto resultTypes = callee.getResultTypes();
+
+    builder.setInsertionPointToEnd(callerEntryBlock);
+    builder.create<CallOp>(caller.getLoc(), calleeName, resultTypes,
+                           fnOperands);
+  }
+
+  /// Conditionally creates an optional new top-level function; and inserts a
+  /// call from the new top-level function to the old top-level function if we
+  /// did create one
+  LogicalResult createOptNewTopLevelFn(ModuleOp moduleOp,
+                                       std::string &topLevelFunction) {
+    auto hasMemrefArguments = [](FuncOp func) {
+      return std::any_of(
+          func.getArguments().begin(), func.getArguments().end(),
+          [](BlockArgument arg) { return isa<MemRefType>(arg.getType()); });
+    };
+
+    /// We only create a new top-level function and call the original top-level
+    /// function from the new one if the original top-level has `memref` in its
+    /// argument
+    bool hasMemrefArgsInTopLevel = false;
+    for (auto funcOp : moduleOp.getOps<FuncOp>()) {
+      if (funcOp.getName() == topLevelFunction) {
+        if (hasMemrefArguments(funcOp)) {
+          hasMemrefArgsInTopLevel = true;
+        }
+      }
+    }
+
+    std::string oldName = topLevelFunction;
+    if (hasMemrefArgsInTopLevel) {
+      auto newTopLevelFunc = createNewTopLevelFn(moduleOp, topLevelFunction);
+
+      OpBuilder builder(moduleOp.getContext());
+      Operation *oldTopLevelFuncOp =
+          SymbolTable::lookupSymbolIn(moduleOp, oldName);
+      auto oldTopLevelFunc = dyn_cast_or_null<FuncOp>(oldTopLevelFuncOp);
+
+      if (!oldTopLevelFunc)
+        oldTopLevelFunc.emitOpError("Original top-level function not found!");
```

Comment:
1. Can we be explicit and change this to `oldTopLevelFunc == nullptr` please? 
2. This should this return right? I don't think you want to pass a nullptr to `insertCallFromNewTopLevel`. Can you add an error test that catches this?

Reply:
I was imagining this case:
```mlir
module {
  func.func @main() : () -> i32 {
     %c0 = constant 0 : i32
     %c1 = constant 1 : i32
     %1 = arith.addi %c0, %c1 : i32
    return %1
  }
  func.func @func_1(%arg0 : i32, %mem0 : memref<8xi32>, %i : index, %mem1 : memref<8xi32>) {
    memref.store %arg0, %mem1[%i] : memref<8xi32>
    memref.store %arg0, %mem0[%i] : memref<8xi32>
    return
  }
}
```
where the `main` function has nothing to do with `func_1` and our command is `lower-scf-to-calyx="top-level-function=main`.

I was wondering, will `func_1` be executed after `main` same as how they are ordered?

Reply:
Perhaps I'm misunderstanding, but one of these would be labeled as the entry point, presumably `main`. If `func_1` is not accessible from the entry point, then it is dead, i.e., it will never be executed. This is not much different than an imperative language today, e.g.,

```cpp
int main() {
  int c0 = 0, c1 = 1;
  return c0 + c1;
  // Since there is no explicit call to `func_1`, it is considered dead code.
} 

void func_1(int arg0, std::vector<int> mem0, size_t i, std::vector<int> mem1) {
  mem1[i] = arg0;
  mem0[i] = arg0;
  return;
}
```

Reply:
I see, then I think I'll just do `assert(oldTopLevelFunc && "Original top-level function not found)` since the `moduleOp` will first try to catch it [here](https://github.com/llvm/circt/blob/17c036f87c4c9e103160b207e18ad595911945e8/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1818) if the top-level function is not set when there are multiple functions. And [this line](https://github.com/llvm/circt/blob/ad89e27ebabb2e50120a3d7468b9e8d738f9c6d5/lib/Conversion/SCFToCalyx/SCFToCalyx.cpp#L1981) will also catch the case when the new top-level is not created properly

So actually couldn't come up the edge case where it will raise `oldTopLevelFunc.emitOpError("Original top-level function not found!")`